### PR TITLE
[#158] Auth middleware

### PR DESCRIPTION
**Pull request by JafarMaash**
# Description:

Resolves issue 158: refactor auth to have middleware
Refactored auth to go through a middleware instead of every endpoint checking the token manually 

auth.js now does what every endpoint used to do on its own. It checks tokens and extracts the user from them, then passes this user onto the controller, to be operated on. 
authentication.js can be named to something like authenticationUtil.js to distinguish since right now there is an "auth" and an "authentication" file. In the interest of getting this pr done asap and avoiding errors due to a renaming, that hasn't been done yet.
Tweaked one test that didn't include auth in its request since it now also goes through middleware. 

As of now a lot of the git diffs in this pr are shuffling around existing LOC, I'll see what I can do about that. 

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing document](../CONTRIBUTING.md)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Refactor (restructured code without changing functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] My code follows the code style of this project
- [ ] My code has been commented
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [x] All new and existing tests passed

### CODE REVIEW

File: backend/middlewares/auth.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import Authentication from './authentication'
+
+module.exports = (req, res, next) => {
```
Comment (scorpionknifes) [inline backend/middlewares/auth.js]: use es6 modules instead of commonjs. e.g. `export` or `export default`

File: backend/middlewares/auth.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+      if (!decodedUser.id) {
+        res.status(401).json({
+          error: new Error('Invalid auth token'),
+        })
+        return
+      } else {
+        res.locals.decodedUser = decodedUser
+        next()
+      }
+    }
```
Comment (scorpionknifes) [inline backend/middlewares/auth.js]:
```suggestion
    const decodedUser = Authentication.extractUser(authToken)
    if (!decodedUser.id) {
      res.status(401).json({
        error: new Error('Invalid auth token'),
      })
      return
    } 
    res.locals.decodedUser = decodedUser
    next()
```

nit: don't need else statement. look cleaner

**Review by scorpionknifes — CHANGES_REQUESTED** [review]
LGTM, couple nits

**Review by DanilLinkov — APPROVED** [review]
LGTM! Nice work

**Review by scorpionknifes — APPROVED** [review]

**Pull request merged**